### DISCUSSION TITLE
Way to configure Consumer to match Producer throughput

### DISCUSSION THREAD

**Discussion by CROprogrammer** [sngular/kloadgen#425] [Q&A]
Hi,

I have a case in which I want to send as many messages from Producer to Topic in a short period, and to read messages from that topic in Consumer. I can achive high throughput in Producer and number of messages I've defined are sent, but Consumer that runs parallel to Producer skips a lot of messages from topic or it does not consume them.

How can I achive that my consumer does not skip messages and that it consumes all messages produced in producer?

When I use Constant Timer then Consumer can process message, but then too few messages can be send to Topic in short period of time.

My test plans:
[Test Plans.zip](https://github.com/sngular/kloadgen/files/13393756/Test.Plans.zip)

```My Avro schema:
{
  "type": "record",
  "namespace": "com.example",
  "name": "Student",
  "fields": [
    { "name": "name", "type": "string" },
    { "name": "age", "type": "int", "default" : -1 }
  ]
}

**Comment by CROprogrammer** [comment]
Does anyone know why this could be?

**Comment by CROprogrammer** [comment]
I've noticed that lag on Kafka Topic is high when sending lots of messages in short period of time, and Consumer at one point stops consuming messages and then a lot of them stay unconsumed and Consumer shuts down. How can Consumer be configured so that no messages become unconsumed?

I would like to send for example a million messages in short period of time, and I would expect for it to be Produced and Consumed in a few seconds.

**Comment by jemacineiras** [comment]
Hi, we need to check this behaviour, Consumer is designed to wait until receive one single message and then deliver to JMeter. An idea we have in mind is to allow sort of "number of messages to consume" so we will consuming messages until this value is reached.
About the lag in Kafka topics... you mean when you produce messages? What kind of setup are you using to run the tests?

**Reply by CROprogrammer** [reply]
So Consumer is designed to consume only one message? Other than that its behaviour is unpredictable?
Lag in Kafka topics I'm talking about is the difference between Producer offset and Consumer offset in particular partition.

I'm using Docker container for Apicurio Schema Registry, in Docker I'm also running a zookeeper with 3 brokers, and Kafdrop to check messages in topics.

**Reply by CROprogrammer** [reply]
My test plans and schema are in original post up there.

**Reply by CROprogrammer** [reply]
For example I'm trying to send 100 000 messages, and Producer does produce them to topic, but Consumer for some reason only consumes let's say about 10% of produced messages and calls it a day.

**Reply by AdrianLagartera** [reply]
> So Consumer is designed to consume only one message? Other than that its behaviour is unpredictable? Lag in Kafka topics I'm talking about is the difference between Producer offset and Consumer offset in particular partition.

Looking at your test plan, I think the problem comes because you are creating 10 producer threads, so I guess that forces Kafka to create more partitions, so since Kafka doesn't distribute the data uniformly, when you have your other 10 consumer threads the partitions produce the "Kafka Lag" trying to fetch the data from many partitions at least this is my guess, because I couldn't test this at all with code. Try to use only one Producer Thread, and keep your 10 Consumer Threads.

**Reply by CROprogrammer** [reply]
I've tried with 1 producer thread and 10 consumer threads, and it works even worse :(
I don't expect number of partitions to be the problem, topics in real projects have more than 1 partition.

**Reply by AdrianLagartera** [reply]
Yeah, I know usually you have more than one partition I was guessing if multiple producers make multiple partitions and that could produce the lag, but yeah, apparently that's not the problem, I will check the properties of your test plan and try to do a load in the same conditions on my local, because I don't remember exactly when, but at some point I did a stress load of a similar amount of messages and it worked fine

**Reply by CROprogrammer** [reply]
Even when there is no lag - assuming that all messages have been consumed, they do not appear in JMeter. I've tried to send 10 000 messages with 1 Producer thread, and cosuming it with 10 Consumer threads, but only 24 messages appeared in JMeter.

![image](https://github.com/sngular/kloadgen/assets/73464772/19c8fe5e-9aaf-43ad-866e-566e51e2be63)

**Reply by CROprogrammer** [reply]
![image](https://github.com/sngular/kloadgen/assets/73464772/1691db66-1229-446e-8a94-05f981ec0929)

![image](https://github.com/sngular/kloadgen/assets/73464772/525c2696-8b8b-4861-847f-c6c09d2fce5b)

**Reply by jemacineiras** [reply]
That is , can you please open an issue so I can check it and see If I should reimplement the consumer?

**Reply by CROprogrammer** [reply]
https://github.com/sngular/kloadgen/issues/429

**Reply by jemacineiras** [reply]
I will be on it asap.
